crates/polars-plan/src/plans/optimizer/set_order.rs

```diff
@@ -733,5 +737,27 @@ pub fn simplify_and_fetch_orderings(
     // Pullup orders from the leaves to the roots.
     pullup_orders(&leaves, ir_arena, &mut outputs, &mut orders, &cache_proxy);
 
-    (orders, cache_proxy)
+    // @Hack. Since not all caches might share the same node and the input of caches might have
+    // been updated, we need to ensure that all caches again have the same input.
+    //
+    // This can be removed when all caches with the same id share the same IR node.
+    for nodes in cache_proxy.into_values() {
+        let updated_node = nodes[0];
```

Comment:
Forgot this, we should also insert `updated_node` into `orders`, it seems that only the first cache will be there.